dashboard/app/controllers/api/v1/ml_models_controller.rb

```diff
@@ -10,8 +10,11 @@ class Api::V1::MlModelsController < Api::V1::JsonApiController
   # Save a trained ML model to S3 and a reference to it in the database.
   def save
     model_id = generate_id
+
+    return head :bad_request if params["ml_model"].nil? || params["ml_model"] == ""
+
     metadata = params["ml_model"].except(:trainedModel, :featureNumberKey)
-    @user_ml_model = UserMlModel.create!(
+    @user_ml_model = UserMlModel.create(
```

Comment:
Why switch from `create!` to `create`? Do we want to catch the exception if there is one?

Reply:
Previously, there wasn't a validation in place for it to raise an exception. I added a name validation in the `UserMlModel` model so I could write a test to mimic the save failing with invalid data. 

With `create!` and the model name validation, the `returns failure when model cannot save` test doesn't succeed because the exception is thrown.

With `create` and the model name validation, it returns a failure response. There's also a name validation in the tool UI; users can't save a model without inputting a name. 

We can also add a client-side validation by specifying `required` in the name form field. I think having UI, database, and client-side validations can make up for having a controller-side validation, which is harder to test. Does that sound okay?

Reply:
Yeah, that sounds ok. Thanks for clarifying.